src/node_file.cc

```diff
@@ -103,6 +104,147 @@ using v8::Value;
+  object()->DefineOwnProperty(env->context(),
+                              FIXED_ONE_BYTE_STRING(env->isolate(), "fd"),
+                              Integer::New(env->isolate(), fd),
+                              attr).FromJust();
+}
+
+FD::~FD() {
+  CHECK(!closing_);  // We should not be deleting while explicitly closing!
+  Close();           // Close synchronously and emit warning
+  CHECK(closed_);    // We have to be closed at the point
+  CHECK(persistent().IsEmpty());
+  if (!object().IsEmpty())
+    ClearWrap(object());
+  persistent().Reset();
+}
+
+
+// Close the file descriptor if it hasn't already been closed. A process
+// warning will be emitted using a SetImmediate to avoid calling back to
+// JS during GC. If closing the fd fails at this point, a fatal exception
+// will crash the process immediately.
+inline void FD::Close() {
+  if (closed_) return;
+  closed_ = true;
+  uv_fs_t req;
+  int ret = uv_fs_close(env()->event_loop(), &req, fd_, nullptr);
+  uv_fs_req_cleanup(&req);
+
+  struct uv_err_detail { int ret; int fd; };
+
+  uv_err_detail* detail = new uv_err_detail { ret, fd_ };
+
+  if (ret < 0) {
+    env()->SetUnrefImmediate([](Environment* env, void* data) {
+      char msg[70];
+      uv_err_detail* detail = static_cast<uv_err_detail*>(data);
+      snprintf(msg, arraysize(msg),
+              "Closing file descriptor %d on garbage collection failed",
+              detail->fd);
+      // This exception will end up being fatal for the process because
+      // it is being thrown from within the SetImmediate handler and
+      // there is no JS stack to bubble it to. In other words, tearing
+      // down the process is the only reasonable thing we can do here.
+      env->ThrowUVException(detail->ret, "close", msg);
+      delete detail;
+    }, detail);
+    return;
+  }
+
+  // If the close was successful, we still want to emit a process warning
+  // to notify that the file descriptor was gc'd. We want to be noisy about
+  // this because not explicitly closing the garbage collector is a bug.
+  env()->SetUnrefImmediate([](Environment* env, void* data) {
+    char msg[70];
+    uv_err_detail* detail = static_cast<uv_err_detail*>(data);
+    snprintf(msg, arraysize(msg),
+            "Closing file descriptor %d on garbage collection",
+            detail->fd);
+    delete detail;
+    ProcessEmitWarning(env, msg);
+  }, detail);
+}
+
+void FD::CloseReq::Resolve() {
+  HandleScope scope(env_->isolate());
+  Local<Promise> promise = promise_.Get(env_->isolate());
+  Local<Promise::Resolver> resolver = promise.As<Promise::Resolver>();
+  resolver->Resolve(env_->context(), Undefined(env_->isolate()));
+}
+
+void FD::CloseReq::Reject(Local<Value> reason) {
+  HandleScope scope(env_->isolate());
+  Local<Promise> promise = promise_.Get(env_->isolate());
+  Local<Promise::Resolver> resolver = promise.As<Promise::Resolver>();
+  resolver->Reject(env_->context(), reason);
```

Comment:
How is async context propagated here? I can’t see anything? (same for resolve)

Reply:
I know less than I need to on that part. What needs to be added here to propagate that?

Reply:
@jasnell Once `CloseReq` is an `ReqWrap`, you’ll want a line like `InternalCallbackScope callback_scope(this);` before the resolve calls. That’s all. :)

(The idea is that the `CloseReq` is an `AsyncWrap` then, so it tracks state automatically, and you only need to enter a scope to make async_hooks listeners aware that things are happening in the `CloseReq`’s async context.)